File: Difficulty: Medium/Rat in a Maze Problem - I/rat-in-a-maze-problem-i.java
//{ Driver Code Starts
// Initial Template for Java

import java.util.*;

class Rat {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();

        while (t-- > 0) {
            int n = sc.nextInt();
            int[][] a = new int[n][n];
            for (int i = 0; i < n; i++)
                for (int j = 0; j < n; j++) a[i][j] = sc.nextInt();

            Solution obj = new Solution();
            ArrayList<String> res = obj.findPath(a);
            Collections.sort(res);
            if (res.size() > 0) {
                for (int i = 0; i < res.size(); i++) System.out.print(res.get(i) + " ");
                System.out.println();
            } else {
                System.out.println(-1);
            }
        }
    }
}

// } Driver Code Ends


class Solution {
    public ArrayList<String> findPath(int[][] mat) {
        // Your code here
        ArrayList<String> res = new ArrayList<>();
        recurse(mat,0,0,"",res);
        return res;
    }
    public static void recurse(int[][] mat, int r, int c, String res, ArrayList<String> ans){
        int rows = mat.length;
        int cols = mat[0].length;
        if(r == rows-1 && c == cols-1 && mat[r][c] == 1){
            ans.add(res);
            return;
        }
        if(r<0 || r>=rows || c<0 || c>=cols){
            return;
        }
        
        if(mat[r][c] == 0 || mat[r][c] == -1){
            return;
        }
        mat[r][c] = -1;
        
        recurse(mat,r-1,c,res+"U",ans);
        recurse(mat,r+1,c,res+"D",ans);
        recurse(mat,r,c-1,res+"L",ans);
        recurse(mat,r,c+1,res+"R",ans);
        
        mat[r][c] = 1;
        
    }
}

File: Difficulty: Medium/Rat in a Maze Problem - I/rat-in-a-maze-problem-i.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function template for C++

class Solution {
  public:
    bool isSafe(int newX, int newY, vector<vector<int>> &vis,vector<vector<int>> &mat, int n){
        if((newX >= 0 && newX < n) && (newY >=0 && newY < n) && vis[newX][newY] == 0 && mat[newX][newY] == 1){
            return true;
        }
        
        return false;
    }
    void solve(int x, int y, vector<vector<int>> &mat, vector<vector<int>> &vis, vector<string> &ans, string path, int n){
        //base case
        if(x == n-1 && y == n-1){
            ans.push_back(path);
            return;
        }
        
        //4 movements -> DLRU
        
        //down
        if(isSafe(x+1,y,vis,mat,n)){
            vis[x][y] = 1;
            solve(x+1,y,mat,vis,ans,path+'D',n);
            vis[x][y] = 0;
        }
        
        //left
        if(isSafe(x,y-1,vis,mat,n)){
            vis[x][y] = 1;
            solve(x,y-1,mat,vis,ans,path+'L',n);
            vis[x][y] = 0;
        }
        
        //right
        if(isSafe(x,y+1,vis,mat,n)){
            vis[x][y] = 1;
            solve(x,y+1,mat,vis,ans,path+'R',n);
            vis[x][y] = 0;
        }
        
        //up
        if(isSafe(x-1,y,vis,mat,n)){
            vis[x][y] = 1;
            solve(x-1,y,mat,vis,ans,path+'U',n);
            vis[x][y] = 0;
        }
    }
    
    vector<string> findPath(vector<vector<int>> &mat) {
        // code here
        vector<string> ans;
        
        int n = mat.size();
        
        if(mat[0][0] == 0){
            return ans;
        }
        
        vector<vector<int>> vis(n,vector<int>(n,0));
        
        solve(0,0,mat,vis,ans,"",n);
        
        return ans;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        string input;
        getline(cin, input);
        vector<vector<int>> mat;
        string innerArray;
        bool isInsideArray = false;

        for (char c : input) {
            if (c == '[') {
                if (isInsideArray) {
                    innerArray.clear();
                }
                isInsideArray = true;
            } else if (c == ']') {
                if (isInsideArray && !innerArray.empty()) {
                    vector<int> row;
                    stringstream ss(innerArray);
                    int num;

                    while (ss >> num) {
                        row.push_back(num);
                        if (ss.peek() == ',')
                            ss.ignore();
                        while (isspace(ss.peek()))
                            ss.ignore();
                    }

                    mat.push_back(row);
                    innerArray.clear();
                }
                isInsideArray = false;
            } else if (isInsideArray) {
                if (!isspace(c)) {
                    innerArray += c;
                }
            }
        }

        Solution obj;
        vector<string> result = obj.findPath(mat);
        sort(result.begin(), result.end());

        if (result.empty())
            cout << "[]";
        else
            for (int i = 0; i < result.size(); i++)
                cout << result[i] << " ";
        cout << endl;
        cout << "~" << endl;
    }
    return 0;
}

// } Driver Code Ends